Peer discovery - maintain k connected peers

solo1g opened the pull request:
This should finish with all aspects of peer discovery. Node can start without any peers in config file except `maxConnectedPeers` (default is 1) and then find, connect and initialize with `maxConnectedPeers` number of peers of the required type (ones supporting compact filters in case of Neutrino) using dns seeds, hardcoded seeds, and addr messages. Of the peers connected, only one will be used for node operation (IBD etc) at the moment.

solo1g committed:
change node init logic to add test connections

Peers are now added first as a test connection and if they are suitable, they are used for node operation else discarded.

solo1g committed:
add peer discovery stack and scheduler

solo1g committed:
fix peer waiting bug

solo1g committed:
change to stack first peer setup

peers are all added to stack first now, from there connections would be handled by scheduler

solo1g committed:
change InitializedDisconnect to disconnect from any state

solo1g committed:
add reconnectSync

node.sync should be called after reconnection iff the reconnected node was one that was used for syncing, else we would have multiple sync calls from different reconnections

solo1g committed:
stop actor when remove unconnected

solo1g committed:
add getaddr message in PeerMessageSender

solo1g committed:
change scheduler params

solo1g committed:
update peers resource file

Update to newer version from https://github.com/bitcoin/bitcoin/blob/master/contrib/seeds/nodes_main.txt

solo1g committed:
fix issue where config peer may not be used

Eg, take 1 config peer and remaining db peers and init all together, it may happen that all from db get connected first. Config peer must take priority in such a case.

solo1g committed:
minor fix

solo1g committed:
fix tests, formatting

solo1g committed:
change messages logged

log level for reconnection related messaged changed to debug, remove log messages added while working

solo1g committed:
test changes

solo1g commented:
Hello @Christewart :wave: . The core stuff is done so wanted an initial review while I fix the tests and ensure that everything is in order.

solo1g committed:
remove server run test

Test was to throw errors if no peers are selected. This test is no longer needed

solo1g committed:
remove akka dead letter logs

Christewart commented:
> Hello @Christewart wave . The core stuff is done so wanted an initial review while I fix the tests and ensure that everything is in order.

Hey Shreyansh! Thanks for this! Unfortunately I don't know how much bandwidth i will have to review this until after Bitcoin 2022 in April. We are trying to get our core wallet software stabilized by then. I will add this to my TODO list to review after that though (or optimistically, we will get all our work done before then! :sweat_smile: )

solo1g committed:
use new peer if current peer keeps disconnecting us

Some peers keep disconnecting us after connection leading to an infinite disconnect reconnect loop. This is different from reconnection limit, as reconnection attempt is successful and one or two header messages too might be sent successfully before a disconnection.

solo1g committed:
rebase

solo1g committed:
fix tests

solo1g committed:
refactor

solo1g committed:
remove unwanted diffs

solo1g committed:
add composite key in PeerTable

solo1g committed:
fix tests

solo1g committed:
change randomPeer to match existing behaviour

solo1g commented:
Alright, so this should be good to go now :face_with_spiral_eyes:.

I wanted to split the whole multi-peer stuff in smaller PRs but often some change would come up that would depend on some other aspect of a multi-peer node outside the scope of the current PR. So you will see a few changes I have made here that do not make sense for a multi-peer neutrino node but are rather temporary "bandages" aka changes to get things working within the context of this PR.
With this said, I'm not sure whether this should be up for merge to master right now or I maintain a separate branch and you review the PRs there built on top of this? What would you prefer? @Christewart 

Oh and as for testing this, add `node.maxConnectedPeers = 8` in your `bitcoin-s.conf` and then if your run without any peers in `node.peers`, this is what you _should_ be seeing
```
2022-06-02T17:06:14UTC INFO [NeutrinoNode] Our node has been full started. It took=244ms
2022-06-02T17:06:14UTC INFO [NeutrinoNode] Started node, best block hash 000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f at height 0, with 0 filter headers and 0 filters
2022-06-02T17:06:14UTC INFO [WalletAppConfig] Starting wallet rebroadcast task
2022-06-02T17:06:14UTC INFO [DLCServer] Bound at /0:0:0:0:0:0:0:0:2862
2022-06-02T17:06:14UTC INFO [PeerManager] Waiting for peer connection
2022-06-02T17:06:14UTC INFO [Server] Started Bitcoin-S websocket at /127.0.0.1:19999
2022-06-02T17:06:14UTC INFO [Server] Started Bitcoin-S HTTP server at /127.0.0.1:9999
2022-06-02T17:06:14UTC INFO [BitcoinSServerMain] Starting neutrino node sync, it took=1813ms
2022-06-02T17:06:24UTC INFO [PeerManager] Connected to peer Peer(46.229.238.187:8333). Connected peer count 1
2022-06-02T17:06:25UTC INFO [PeerManager] Connected to peer. Starting sync.
2022-06-02T17:06:25UTC INFO [NeutrinoNode] Starting sync node, height=0 hash=000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f
2022-06-02T17:06:25UTC INFO [BitcoinSServerMain] Done starting Main! It took 12408ms
2022-06-02T17:06:25UTC INFO [BitcoinSServerMain] Done start BitcoinSServerMain, it took=14319ms
2022-06-02T17:06:26UTC INFO [DataMessageHandler] Received headers message with 2000 headers
2022-06-02T17:06:27UTC INFO [ChainHandler] Processed headers from height=0 to 2000. Best hash=00000000dfd5d65c9d8561b4b8f60a63018fe3933ecb131fb37f905f87da951a
2022-06-02T17:06:27UTC INFO [DataMessageHandler] Received maximum amount of headers in one header message. This means we are not synced, requesting more
2022-06-02T17:06:28UTC INFO [DataMessageHandler] Received headers message with 2000 headers
2022-06-02T17:06:29UTC INFO [ChainHandler] Processed headers from height=2000 to 4000. Best hash=00000000922e2aa9e84a474350a3555f49f06061fd49df50a9352f156692a842
2022-06-02T17:06:29UTC INFO [DataMessageHandler] Received maximum amount of headers in one header message. This means we are not synced, requesting more
2022-06-02T17:06:29UTC INFO [DataMessageHandler] Received headers message with 2000 headers
2022-06-02T17:06:30UTC INFO [ChainHandler] Processed headers from height=4000 to 6000. Best hash=00000000dbbb79792303bdd1c6c4d7ab9c21bba0667213c2eca955e11230c5a5
2022-06-02T17:06:30UTC INFO [DataMessageHandler] Received maximum amount of headers in one header message. This means we are not synced, requesting more
2022-06-02T17:06:31UTC INFO [DataMessageHandler] Received headers message with 2000 headers
2022-06-02T17:06:32UTC INFO [ChainHandler] Processed headers from height=6000 to 8000. Best hash=0000000094fbacdffec05aea9847000522a258c269ae37a74a818afb96fc27d9
2022-06-02T17:06:32UTC INFO [DataMessageHandler] Received maximum amount of headers in one header message. This means we are not synced, requesting more
2022-06-02T17:06:32UTC INFO [DataMessageHandler] Received headers message with 2000 headers
2022-06-02T17:06:33UTC INFO [ChainHandler] Processed headers from height=8000 to 10000. Best hash=0000000099c744455f58e6c6e98b671e1bf7f37346bfd4cf5d0274ad8ee660cb
2022-06-02T17:06:33UTC INFO [DataMessageHandler] Received maximum amount of headers in one header message. This means we are not synced, requesting more
2022-06-02T17:06:34UTC INFO [PeerManager] Connected to peer Peer(18.181.194.174:8333). Connected peer count 2
2022-06-02T17:06:34UTC INFO [DataMessageHandler] Received headers message with 2000 headers
2022-06-02T17:06:34UTC INFO [PeerManager] Connected to peer Peer(96.58.134.226:8333). Connected peer count 3
2022-06-02T17:06:35UTC INFO [ChainHandler] Processed headers from height=10000 to 12000. Best hash=0000000011d1d9f1af3e1d038cebba251f933102dbe181d46a7966191b3299ee
2022-06-02T17:06:35UTC INFO [DataMessageHandler] Received maximum amount of headers in one header message. This means we are not synced, requesting more
2022-06-02T17:06:36UTC INFO [DataMessageHandler] Received headers message with 2000 headers
2022-06-02T17:06:37UTC INFO [ChainHandler] Processed headers from height=12000 to 14000. Best hash=000000002d9050318ec8112057423e30b9570b39998aacd00ca648216525fce3
2022-06-02T17:06:37UTC INFO [DataMessageHandler] Received maximum amount of headers in one header message. This means we are not synced, requesting more
2022-06-02T17:06:38UTC INFO [DataMessageHandler] Received headers message with 2000 headers
2022-06-02T17:06:39UTC INFO [ChainHandler] Processed headers from height=14000 to 16000. Best hash=00000000679a1ab3af6da03f13a0bc96d7215e65458b2d2edfa030b5b431e8b3
2022-06-02T17:06:39UTC INFO [DataMessageHandler] Received maximum amount of headers in one header message. This means we are not synced, requesting more
2022-06-02T17:06:40UTC INFO [DataMessageHandler] Received headers message with 2000 headers
2022-06-02T17:06:41UTC INFO [ChainHandler] Processed headers from height=16000 to 18000. Best hash=00000000f914f0d0692e56bd06565ac4de668251b6a29fe0535d1e0031cfd0de
2022-06-02T17:06:41UTC INFO [DataMessageHandler] Received maximum amount of headers in one header message. This means we are not synced, requesting more
2022-06-02T17:06:42UTC INFO [DataMessageHandler] Received headers message with 2000 headers
2022-06-02T17:06:42UTC INFO [ChainHandler] Processed headers from height=18000 to 20000. Best hash=00000000770ebe897270ca5f6d539d8afb4ea4f4e757761a34ca82e17207d886
2022-06-02T17:06:42UTC INFO [DataMessageHandler] Received maximum amount of headers in one header message. This means we are not synced, requesting more
2022-06-02T17:06:43UTC INFO [DataMessageHandler] Received headers message with 2000 headers
2022-06-02T17:06:44UTC INFO [PeerManager] Connected to peer Peer(136.243.137.166:8333). Connected peer count 4
2022-06-02T17:06:44UTC INFO [ChainHandler] Processed headers from height=20000 to 22000. Best hash=000000004625a14242beccb38c63a1f770a76ee5788764e6c0abd4129bbc1b9d
2022-06-02T17:06:44UTC INFO [DataMessageHandler] Received maximum amount of headers in one header message. This means we are not synced, requesting more
2022-06-02T17:06:44UTC INFO [PeerManager] Connected to peer Peer(3.235.183.74:8333). Connected peer count 5
2022-06-02T17:06:45UTC INFO [DataMessageHandler] Received headers message with 2000 headers
2022-06-02T17:06:46UTC INFO [ChainHandler] Processed headers from height=22000 to 24000. Best hash=00000000f04fccc81f37002707e9501a3f7bdcf25f65531f386a2da8af20122e
2022-06-02T17:06:46UTC INFO [DataMessageHandler] Received maximum amount of headers in one header message. This means we are not synced, requesting more
2022-06-02T17:06:47UTC INFO [DataMessageHandler] Received headers message with 2000 headers
2022-06-02T17:06:48UTC INFO [ChainHandler] Processed headers from height=24000 to 26000. Best hash=000000006d6c151db6d4d67356d590a897a11cd7d8111ee989de6f2f548410bf
2022-06-02T17:06:48UTC INFO [DataMessageHandler] Received maximum amount of headers in one header message. This means we are not synced, requesting more
2022-06-02T17:06:49UTC INFO [DataMessageHandler] Received headers message with 2000 headers
2022-06-02T17:06:50UTC INFO [ChainHandler] Processed headers from height=26000 to 28000. Best hash=00000000172c5ed49d7dfc29bf9a18a53fa2d050fa37aa210d6d4080fd0c7e67
2022-06-02T17:06:50UTC INFO [DataMessageHandler] Received maximum amount of headers in one header message. This means we are not synced, requesting more
2022-06-02T17:06:51UTC INFO [DataMessageHandler] Received headers message with 2000 headers
2022-06-02T17:06:52UTC INFO [ChainHandler] Processed headers from height=28000 to 30000. Best hash=00000000de1250dc2df5cf4d877e055f338d6ed1ab504d5b71c097cdccd00e13
2022-06-02T17:06:52UTC INFO [DataMessageHandler] Received maximum amount of headers in one header message. This means we are not synced, requesting more
2022-06-02T17:06:53UTC INFO [DataMessageHandler] Received headers message with 2000 headers
2022-06-02T17:06:54UTC INFO [ChainHandler] Processed headers from height=30000 to 32000. Best hash=00000000049172ba3ec1b673cf13e3d0049c1c07bb103ed3fa300e3833480055
2022-06-02T17:06:54UTC INFO [DataMessageHandler] Received maximum amount of headers in one header message. This means we are not synced, requesting more
2022-06-02T17:06:54UTC INFO [PeerManager] Connected to peer Peer(70.35.98.12:8333). Connected peer count 6
2022-06-02T17:06:55UTC INFO [DataMessageHandler] Received headers message with 2000 headers
```
All these are peers with compact filters.

solo1g committed:
add back NeutrinoUnsupportedPeerTest

solo1g committed:
connect and store all peers
